frontend/test/lit/test_peephole_optimizations.py

```diff
@@ -0,0 +1,114 @@
+Each test has two components:
+  1. A qnode with a peephole optimization applied, usually called "f"
+  2. The SAME qnode without a peephole optimization applied, usually called "g"
+
+We need to check that:
+  1. For "f", the peephole transform is correctly applied in mlir
+  2. For "g", the peephole transform is correctly not applied in mlir
+  3. "f" and "g" returns the same results.
+"""
+
+# RUN: %PYTHON %s | FileCheck %s
+
+# pylint: disable=line-too-long
+
+import shutil
+
+import numpy as np
+import pennylane as qml
+
+from catalyst import cancel_inverses, qjit
+
+
+def flush_peephole_opted_mlir_to_iostream(filename):
+    """
+    The QJIT compiler does not offer an interface to access an intermediate mlir in the pipeline.
+    The `QJIT.mlir` is the mlir before any passes are run, i.e. the "0_<qnode_name>.mlir".
+    Since the QUANTUM_COMPILATION_PASS is located in the middle of the pipeline, we need
+    to retrieve it with keep_intermediate=True and manually access the "2_QuantumCompilationPass.mlir".
+    Then we delete the kept intermediates to avoid pollution of the workspace
+    """
+    with open(filename + "/2_QuantumCompilationPass.mlir", encoding="utf-8") as file:
+        print(file.read())
+    shutil.rmtree(filename)
+
+
+# CHECK-LABEL: public @jit_cancel_inverses_not_applied
+@qjit(keep_intermediate=True)
+def cancel_inverses_not_applied(xx: float):
+    """
+    Test that a qnode without catalyst.cancel_inverses produces the expected gates in the mlir.
+    """
+
+    @qml.qnode(qml.device("lightning.qubit", wires=1))
+    def f(x: float):
+        qml.RX(x, wires=0)
+        qml.Hadamard(wires=0)
+        qml.Hadamard(wires=0)
+        return qml.expval(qml.PauliY(0))
+
+    # CHECK: {{%.+}} = quantum.custom "RX"({{%.+}}) {{%.+}} : !quantum.bit
+    # CHECK: {{%.+}} = quantum.custom "Hadamard"() {{%.+}} : !quantum.bit
+    # CHECK: {{%.+}} = quantum.custom "Hadamard"() {{%.+}} : !quantum.bit
+    _ff = f(xx)
+
+    return _ff
+
+
+flush_peephole_opted_mlir_to_iostream("cancel_inverses_not_applied")
+
+
+# CHECK-LABEL: public @jit_cancel_inverses_workflow
+@qjit(keep_intermediate=True)
+def cancel_inverses_workflow(xx: float):
+    """
+    Test catalyst.cancel_inverses
+    """
+
+    @cancel_inverses
+    @qml.qnode(qml.device("lightning.qubit", wires=1))
+    def f(x: float):
+        qml.RX(x, wires=0)
+        qml.Hadamard(wires=0)
+        qml.Hadamard(wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+    @qml.qnode(qml.device("lightning.qubit", wires=1))
+    def g(x: float):
+        qml.RX(x, wires=0)
+        qml.Hadamard(wires=0)
+        qml.Hadamard(wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+    # CHECK: {{%.+}} = quantum.custom "RX"({{%.+}}) {{%.+}} : !quantum.bit
+    # CHECK-NOT: {{%.+}} = quantum.custom "Hadamard"() {{%.+}} : !quantum.bit
+    # CHECK-NOT: {{%.+}} = quantum.custom "Hadamard"() {{%.+}} : !quantum.bit
```

Comment:
While the second `CHECK-NOT` is unnecessary (if something is not there, then it's also "not there twice"), it emphasizes the fact that both Hadamards need to be removed. 

Thoughts on keeping/removing this second `CHECK-NOT`?

Reply:
You can also use CHECK-NEXT to ensure that the next instruction is the other RX. And CHECK-NOTs around the CHECK-NEXTs. This would ensure that the circuit is exactly what you expect and that there are no more quantum instructions outside of what you expect.

Reply:
done